Restore built-in FluidSynth effects

johnnovak opened the pull request:
In the SF2 SoundFont format, different chorus and reverb settings can be assigned to individual instruments layers. The creators of such SoundFonts can (and do) tweak the effect levels of different instruments to achieve the desired outcome. Therefore, removing the FluidSynth chorus and reverb and replacing them with global effect processing at the audio channel level is just not feasible.

I've tried a few different SF2 files with the new global processing, and compared them to the old behaviour, and unsurprisingly, everything sounds way off with the global effects. It's just not possible to find a single global preset that would work well — that's easy to understand conceptually, e.g. some instruments in any given SF2 file might have 0% reverb applied to them, while some other instruments might have 100% reverb, etc. Apart from the samples used, the per-instrument chorus & reverb choices give all the different SoundFonts their sonic character — we definitely don't want to alter or remove that!

The situation is similar to the MT-32 emulation; there the reverb & chorus levels are being controlled dynamically by the MIDI data, therefore we can't just replace those effects with some fixed global processing either.

Really, FluidSynth and MT-32 must be treated as hardware-synthesizers that do their own instrument-level effect processing and/or dynamic processing based on MIDI data, therefore mixer level effects should always be disabled for them.

Sorry, @kcgen that I missed this when you raised your first PR for removing the FluidSynth reverb; somehow I forgot about this...

kcgen commented:
Trying to get a handle on context..

The SF2 specification came out in 1998, essentially post-dating the DOS General MIDI era (http://www.synthfont.com/SFSPEC21.PDF); which explains why mid-190' DOS-era soundfont rips sound rather antiseptic and flat -- because the hardware didn't have these effect-filters available.

That said, we certainly modern-era soundfont instruments to be constructed with whatever internal effects are specified by the SF2 author (and it makes sense, a piano's size and material will give it it's own inherent reverb and chorus behavior).

At the same time, at the game-level, users might still want ambient reverb for the "space" in which the music is played, which I believe is handled by the mixer-level reverb and chorus, correct?

johnnovak commented:
Good job unearthing those specs, @kcgen, I've learned a few new things about the GM standard!

In a nutshell, you've convinced me that the user should have the option to turn off the onboard FluidSynth reverb completely, and opt for the new mixer-level reverb instead. The jury is still out whether it is desirable to add the two on top of each other, but yeah, by giving the user complete control over the two reverbs, this would be a possibility.

In the spirit of furthering our knowledge about all things DOS-audio, here's a few remarks.

> The SF2 specification came out in 1998, essentially post-dating the DOS General MIDI era (http://www.synthfont.com/SFSPEC21.PDF);

It was actually introduced much earlier. From [Wikipedia](https://en.wikipedia.org/wiki/SoundFont):

"The original SoundFont file format was developed in the early 1990s by E-mu Systems and Creative Labs. A specification for this version was never released to the public. **The first and only major device to utilize this version was Creative's Sound Blaster AWE32 in 1994.**"

"**SoundFont 2.0 was developed in 1996.** This file format generalized the data representation using perceptually additive real world units, redefined some of the instrument layering features within the format, added true stereo sample support and removed some obscure features of the first version whose behavior was difficult to specify."

So we can conclude that the SoundFont format was effectively introduced to the DOS-gaming world in 1994 with the release of the AWE32.

> which explains why mid-90's DOS-era soundfont rips sound rather antiseptic and flat -- because the hardware didn't have these effect-filters available.

It doesn't quite explain it the way you're suggesting, as the AWE32 has onboard reverb and chorus, so those effects were available for SoundFont authors from the moment the SF2 format became available. I think it's more down to individual taste — some SoundFont authors just don't like reverb, apparently.

From [Wikipedia](https://en.wikipedia.org/wiki/Sound_Blaster_AWE32) again:

"The synthesizer section consisted of the EMU8000 synthesizer and effects processor chip, 1 MB EMU8011 sample ROM, and a variable amount of RAM (none on the SB32, 512 KB on the AWE32; RAM was expandable to 28 MB on both cards). These chips comprised a powerful and flexible sample-based synthesis system, based on E-mu's high-end sampler systems such as the E-mu Emulator III and E-mu Proteus. **The effects processor generated various effects (i.e. reverb and chorus) and environments on MIDI output, similar to the later EAX standard on Live! and newer cards.** It can also add effects to the output from the Yamaha OPL3's FM synthesis. **The AWE32 was the first sampler to support E-Mu's SoundFont standard**, which allowed users to build custom sound sets using their own samples, the samples included in ROM, or both. **The card was sold with software for building custom SoundFonts.** All of Creative's subsequent cards, other than the Sound Blaster PCI64/128 series, support SoundFonts. "

So it's pretty clear from this that the SoundFont format started to conquer the world from 1994 onwards, with reverb and chorus support from the moment it was introduced.

------

So let's have a look at [Creative_1mgm.sf2](https://archive.org/download/free-soundfonts-sf2-2019-04/Creative_1mgm.sf2), which is supposed to be the official SoundFont 2 version of the original AWE32 ROM soundset released later by Emu/Creative Labs themselves (btw, you can find all important DOS-gaming related SF2 files on [this page](https://archive.org/download/free-soundfonts-sf2-2019-04)). I loaded it into the [Polyphone](https://www.polyphone-soundfonts.com/) SF2 editor, and as you can see, while the reverb is fixed on all instruments in this particular soundfont, the chorus settings vary per instrument (chorus & reverb send percentages ordered by descending order):

![image](https://user-images.githubusercontent.com/698770/180671396-abd9dd45-d359-408d-a825-9ead1e91d722.png)

![image](https://user-images.githubusercontent.com/698770/180671436-541cb1c9-db96-41ee-a3a9-4faae0025bbc.png)

But then some SF2 files are authored quite dry and upfront, I think that's what you're referring to. But that was an authoring choice, they could have cranked up the reverb to 100 on all instruments if they wanted. This is from Creative's own [Creative Labs 2M GM_2gmgsmt.sf2](https://archive.org/download/free-soundfonts-sf2-2019-04/Creative%20Labs%202M%20GM_2gmgsmt.sf2):

<img width="1059" alt="image" src="https://user-images.githubusercontent.com/698770/180673284-9607d8d4-3fac-4434-9851-5dfbf4781336.png">

Okay, let's look at something more modern, the popular [Timbres Of Heaven GM_GS_XG_SFX V 3.4 Final.sf2](https://archive.org/download/free-soundfonts-sf2-2019-04/Timbres%20Of%20Heaven%20GM_GS_XG_SFX%20V%203.4%20Final.sf2):

<img width="1345" alt="image" src="https://user-images.githubusercontent.com/698770/180673762-6cc66c2b-934f-44fb-b7c3-c83ae6f615a2.png">

Now we're talking, there's a good amount of reverb on many of the instruments. If you scroll down to the bottom, some instrument have little or even zero reverb. So clearly, the author made some decisions on what should sound upfront and dry, and what should be drenched in reverb.

A popular SC-55 recreation just uses 50% for everything:

![image](https://user-images.githubusercontent.com/698770/180674026-93e453a6-f5c6-4ec3-91de-aebb58a2a215.png)

I could go on, but it's easy to see a pattern — it's all over the place.

It's also important to realise that the SF2 file only specifies per-instrument reverb and chorus send levels as percentages, but says _absolutely nothing_ about the specifics of the reverb itself — not even its relative volume in relation to the samples. I guess the assumption is to use "sensible" levels, comparable to the level of the samples, then the user could adjust the reverb room size preset according to taste. So by choosing a short room, medium room, or large cathedral room size preset in the SF2 player itself, or adjusting the master level reverb volume, one could coax quite different sounds out of the exact same SF2 file. Our default FluidSynth settings just happen to be some kind of middle-of-the-road settings we've taked fromt the ScummVM project — probably not particularly suited to any SF2 file, but sounds OK with most of them, but probably not spectacular with any of them.

-----

So given all of the above, by enabling `ChannelFeature::ReverbSend` for the `FSYNTH` channel, the user could configure all possible combinations:

1. Disable the new global reverb and only use the FluidSynth reverb to get the per-instrument reverb & chorus variations as the author intended (our old behaviour).
3. Enable the new global reverb and disable the FluidSynth reverb. One would miss out on the per-instrument variations this way, but okay, if some people really wanted this, why not give them the option.
4. Enable both reverbs at the same time. Depending on the SF2 file and the configuration of both reverbs, this might sound fine, even good, or like complete crap (e.g. like listening to some reverb-heavy music through a loudspeaker in a cave...)
5. Disable both the new reverb and the FluidSynth reverb to get a completely dry sound (not sure if many people would want this).

I still think that 1) is the best option for SoundFonts, but the current very generic reverb & chorus presets taken from ScummVM sound a bit anemic with many SF2 files, I agree. The current full ability to fully customise the FS reverb & chorus is good, but I think we also need to introduce presets, similar to the global reverb. Then one could simply just say `fsynth_reverb = large 150` to set the "large" FluidSynth reverb preset at 150% volume level.

So... given all of the above, what I'm proposing is to enable global reverb sends to the FluidSynth channel. The FluidSynth reverb would be restored and on by default, and the new global reverb off — existing users wouldn't notice any change in behaviour. I'll include a short notice in the description of the `reverb` and `fsynth_reverb` params to be careful about "stacking" these effects (same for the chorus params). Then eventually (not in 0.79, but later) we could introduce a preset system the FluidSynth effects as well.

How does this sound?

kcgen commented:
Sounds great - and thank you for the clarifications regarding SF2 spec and how DOS era GM (1.0-spec) hardware /did/ have access to reverb and chorus! 

Given the hardware was available, it's actually unfortunate the GM 1.0 spec didn't include handling for system-level reverb, while no doubt lead to some SF2 authors over-soaking their instruments with reverb to try to compensate. Very interesting stuff!

Given the SF2's are all over the board, your proposal to allow toggling both the SF2-specific reverb and system-level reverb separately allows the user the flexibility to handle the range of SF2s.

> And for completeness' sake, I still want to leave MT-32 completely alone — no global reverb & chorus sends for that channel.

Makes sense! Sound like an ideal way forward - thanks for sharing all your findings!

johnnovak commented:
> With SF2's all over the board reverb-wise, I fully agree with your proposal to allow toggling both the SF2-specific reverb level (via FluidSynth) and the mixer-level reverb, given it's the only way to handle the combinations you've nicely described.

Yeah, and we need to make sure to document it adequately but succinctly so non-technical users who are disinterested by all this historical context can get the most out of their SF2 files. ...in the new upcoming documentation project 🎉

johnnovak commented:
Made the changes as discussed @kcgen.

kcgen commented:
Looks good, @johnnovak !
